cf-oscore/src/main/java/org/eclipse/californium/oscore/Encryptor.java

```diff
@@ -160,17 +160,30 @@ protected static Encrypt0Message prepareCOSEStructure(byte[] confidential) {
 		int firstByte = 0x00;
 		ByteArrayOutputStream bRes = new ByteArrayOutputStream();
 		byte[] partialIV = OSSerializer.processPartialIV(ctx.getSenderSeq());
-		firstByte = firstByte | (partialIV.length & 0x07);
-		firstByte = firstByte | 0x08;
+		firstByte = firstByte | (partialIV.length & 0x07); //PartialIV length
+		firstByte = firstByte | 0x08; //Set the KID bit
+
+		//If the Context ID should be included for this context, set its bit
+		if(ctx.getIncludeContextId() == true) {
+			firstByte = firstByte | 0x10;
+		}
 
 		bRes.write(firstByte);
+
 		try {
 			bRes.write(partialIV);
+
+			//Encode the Context ID length and value if to be included
+			if(ctx.getIncludeContextId() == true) {
+				bRes.write(ctx.getIdContext().length);
```

Comment:
In the constructor of `OSCoreCtx` the contextId can be `null`.

```
if (contextId != null) {
	this.context_id = contextId.clone();
} else {
	this.context_id = null;
}
```

Is there any mechanism, that ensures, `ctx.getIdContext().length` doesn't throw a `NullPointerException`?

Reply:
The idea I had was that the developer should never use setIncludeContextId(true) on a context that does not actually have a context ID.

If this is not sufficient I can change the following 4 instances in Encryptor.java:
`if (ctx.getIncludeContextId()) {`
to
`if (ctx.getIncludeContextId() && ctx.getIdContext() != null) {`

Reply:
> The idea I had was that the developer should never use setIncludeContextId(true) on a context that does not actually have a context ID.

Consider to mention that at least in the javadoc.
If it's ensured, that the context is already set before the `setIncludeContextId(true)`, then a check and "IllegalStateException" on failure would be preferred.

Reply:
I have now added some code to handle this in commit 1ddce35dad504a0eff94b5001d128ed04aa6dfc7

Is this like what you had in mind?

Reply:
At least, very close :-).

In the most cases, when a `RuntimeException` is thrown in californium, that's documented in the javadoc. e.g. 

https://github.com/eclipse/californium/blob/ed6a88f0490218ac4928ba1800c11aa806cf490a/element-connector/src/main/java/org/eclipse/californium/elements/RawData.java#L141

Would it be possible, that you add the `IllegalStateException` to javadoc?

Reply:
Yes, I should have added it to the Javadoc too. I have now fixed this in commit 792cfea3fff64e759eb5bcbbdc35ccf6934dc540